analytic_engine/src/sst/parquet/writer.rs

```diff
@@ -149,75 +128,131 @@ impl RecordBytesReader {
+        let mut total_num_rows = 0;
 
         let mut parquet_encoder = ParquetEncoder::try_new(
+            sink,
+            &self.meta_data.schema,
             self.hybrid_encoding,
             self.num_rows_per_row_group,
+            self.max_buffer_size,
             self.compression,
-            parquet_meta_data,
         )
         .box_err()
         .context(EncodeRecordBatch)?;
 
-        // process record batch stream
-        let mut arrow_record_batch_vec = Vec::new();
-        for record_batches in self.partitioned_record_batch {
-            for batch in record_batches {
-                arrow_record_batch_vec.push(batch.into_record_batch().into_arrow_record_batch());
+        loop {
+            let row_group = self.fetch_next_row_group(&mut prev_record_batch).await?;
+            if row_group.is_empty() {
+                break;
             }
 
-            let buf_len = arrow_record_batch_vec.len();
-            let row_num = parquet_encoder
-                .encode_record_batch(arrow_record_batch_vec)
+            self.build_parquet_filter(&row_group, &mut parquet_filter)?;
+
+            let num_batches = row_group.len();
+            for record_batch in row_group {
+                arrow_row_group.push(record_batch.into_record_batch().into_arrow_record_batch());
+            }
+            let num_rows = parquet_encoder
+                .encode_record_batches(arrow_row_group)
+                .await
                 .box_err()
                 .context(EncodeRecordBatch)?;
-            self.total_row_num.fetch_add(row_num, Ordering::Relaxed);
-            arrow_record_batch_vec = Vec::with_capacity(buf_len);
+
+            // TODO: it will be better to use `arrow_row_group.clear()` to reuse the
+            // allocated memory.
+            arrow_row_group = Vec::with_capacity(num_batches);
+            total_num_rows += num_rows;
         }
 
-        let bytes = parquet_encoder
+        let parquet_meta_data = {
+            let mut parquet_meta_data = ParquetMetaData::from(self.meta_data);
+            parquet_meta_data.parquet_filter = Some(parquet_filter);
+            parquet_meta_data
+        };
+        parquet_encoder
+            .set_meta_data(parquet_meta_data)
+            .box_err()
+            .context(EncodeRecordBatch)?;
+
+        parquet_encoder
             .close()
+            .await
             .box_err()
             .context(EncodeRecordBatch)?;
-        Ok(bytes)
+
+        Ok(total_num_rows)
+    }
+}
+
+struct ObjectStoreMultiUploadAborter<'a> {
+    location: &'a Path,
+    sink_id: String,
+    object_store: &'a ObjectStoreRef,
+}
+
+impl<'a> ObjectStoreMultiUploadAborter<'a> {
+    async fn initialize_upload(
+        object_store: &'a ObjectStoreRef,
+        location: &'a Path,
+    ) -> Result<(
+        ObjectStoreMultiUploadAborter<'a>,
+        Box<dyn AsyncWrite + Unpin + Send>,
+    )> {
+        let (sink_id, sink) = object_store
```

Comment:
this module never has `sink` variable, I prefer to name it `writer`.

Reply:
How about `upload_writer`?